src/engraving/rendering/dev/layoutindependent.cpp

```diff
@@ -0,0 +1,237 @@
+        case AccidentalBracket::BRACE: return { SymId::accidentalCombiningOpenCurlyBrace, SymId::accidentalCombiningCloseCurlyBrace };
+        case AccidentalBracket::NONE: return { SymId::noSym, SymId::noSym };
+        }
+        return { SymId::noSym, SymId::noSym };
+    };
+
+    // Single?
+    SymId singleSym = accidentalSingleSym(item);
+    if (singleSym != SymId::noSym && conf.engravingFont()->isValid(singleSym)) {
+        Accidental::LayoutData::Sym s(singleSym, 0.0, 0.0);
+        ldata->syms.push_back(s);
+
+        ldata->addBbox(item->symBbox(singleSym));
+    }
+    // Multi
+    else {
+        double margin = conf.styleMM(Sid::bracketedAccidentalPadding);
+        double x = 0.0;
+
+        std::pair<SymId, SymId> bracketSyms;
+        bool isNeedBracket = item->bracket() != AccidentalBracket::NONE && !item->parentNoteHasParentheses();
+        if (isNeedBracket) {
+            bracketSyms = accidentalBracketSyms(item->bracket());
+        }
+
+        // Left
+        if (bracketSyms.first != SymId::noSym) {
+            Accidental::LayoutData::Sym ls(bracketSyms.first, 0.0,
+                                           item->bracket() == AccidentalBracket::BRACE ? item->spatium() * 0.4 : 0.0);
+            ldata->syms.push_back(ls);
+            ldata->addBbox(item->symBbox(bracketSyms.first));
+
+            x += item->symAdvance(bracketSyms.first) + margin;
+        }
+
+        // Main
+        SymId mainSym = item->symId();
+        Accidental::LayoutData::Sym ms(mainSym, x, 0.0);
+        ldata->syms.push_back(ms);
+        ldata->addBbox(item->symBbox(mainSym).translated(x, 0.0));
+
+        // Right
+        if (bracketSyms.second != SymId::noSym) {
+            x += item->symAdvance(mainSym) + margin;
+
+            Accidental::LayoutData::Sym rs(bracketSyms.second, x,
+                                           item->bracket() == AccidentalBracket::BRACE ? item->spatium() * 0.4 : 0.0);
+            ldata->syms.push_back(rs);
+            ldata->addBbox(item->symBbox(bracketSyms.second).translated(x, 0.0));
+        }
+    }
+}
+
+void LayoutIndependent::layout(const ActionIcon* item, ActionIcon::LayoutData* ldata)
+{
+    FontMetrics fontMetrics(item->iconFont());
+    ldata->setBbox(fontMetrics.boundingRect(Char(item->icon())));
+}
+
+void LayoutIndependent::layout(const Articulation* item, Articulation::LayoutData* ldata)
+{
+    if (item->isHiddenOnTabStaff()) {
+        ldata->setIsSkipDraw(true);
+        return;
+    }
+    ldata->setIsSkipDraw(false);
+
+    RectF bbox;
+
+    if (item->textType() == ArticulationTextType::NO_TEXT) {
+        bbox = item->symBbox(item->symId());
+
+        //! NOTE symId can be changed during layout if anchor is auto
+        static mu::async::Asyncable holder;
+        item->symIdChanged().resetOnReceive(&holder);
+        item->symIdChanged().onReceive(&holder, [item, ldata](SymId val) {
```

Comment:
[reposting this comment because it was lost]

Actually, I don't really like using Channel/Asyncable here. For two reasons:
- it makes the code path complicated. Control flow will jump all over the place, and it's difficult to keep track of what will happen by calling a certain function, because you need to find out which Channels and Notifications will be triggered. 
- it does not solve the core problem. The core problem is: we need to calculate things in the correct order. We wish to calculate layout data for "leaf" items all at once, because it seems easy. But this means:
    - we are doing the initial calculation based on incomplete data (because the definitive symId is not yet calculated)
    - every time that some part of the data that we depend on is calculated (i.e. the symId gets calculated), we need to redo the bbox calculation of the Articulation. So we calculate the same thing multiple times, while only the last time makes sense.

    In the case of just Articulation, there is only one property on which the layout data directly depends, namely the symId, the problem is maybe not so big (the non-definitive symId is not harmful, and the amount of recalculation won't be huge), but I don't think this is the way to go when moving forward.

    Conclusion, even though it seems a nice wish to layout leaf items all at once, it turns out that it is no good idea: those leaf items _do_ depend on their parent and context. So, first the required surrounding data needs to be calculated before we can calculate data about these items.

Reply:
It all looks like the pass system is not suitable, because every time we will come across the fact that some dependency has not yet been calculated. And then we need to do either undeterminated passes (without responsibility, but just pieces of one algorithm), or we need to do hundreds of small passes, and perhaps make a graph from these passes :)

Reply:
It looks good to me now. Is the plan to merge this right now, and then still look into dependency graphs instead of passes in a next PR? Or are we still going ahead with the idea of passes?

Reply:
I would prefer merge